client/src/graphql/hooks/useQueryWithErrorHandling.ts

```
@@ -0,0 +1,56 @@
+import {
+  useQuery,
+  ApolloQueryResult,
+  OperationVariables,
+  DocumentNode,
+  QueryHookOptions,
+} from "@apollo/client";
+
+/**
+ * Throw the various errors that might come from a GraphQL query response.
+ */
+export const throwGqlError = <T, U extends ApolloQueryResult<T | undefined>>(
+  result: U,
+  key: keyof T
+) => {
+  if (result.loading) {
+    return result;
+  }
+
+  if (result.error) {
+    throw result.error;
+  }
+
+  if (!result.data) {
+    throw new Error("GQL_MISSING_DATA");
+  }
+
+  if (!result.data[key]) {
+    throw new Error("GQL_MISSING_DATA_KEY");
+  }
+
+  // TODO: decide whether to throw these or just log them.
+  result.errors?.forEach((e) => {
+    console.warn(e);
+  });
+
+  return result;
+};
+
+/**
+ * Variation of apollo's useQuery hook with automatic error handling.
+ *
+ * Errors get thrown, presumably will be caught by an ErrorBoundary in the
```

Can we add a test to confirm this?

---

maybe i could rephrase the comment, the point is just to say that errors are not handled directly within this component and it should be used with an ErrorBoundary (this is basically the same advice react prints out in the console whenever it sees an unhandled error). if we wanted to test we're using an ErrorBoundary and it's handling errors, we'd have to write an integration test.

---

Right, we'd have to simulate an error somehow for the actual error boundary. Was it working if you manually tested it? I'm assuming it came up in dev and having the error boundary was helpful to catch stack traces. 

I had saved this video for eventual implementation of an error boundary! https://www.youtube.com/watch?v=dxWrHEOD5DU

---

yea, i tested manually.

i'd think of the error boundary as similar to a try-catch block. we don't need to test that the error boundary catches errors, we can trust that it does (it's part of the react framework). but in the same way with try-catch, we _do_ want to make sure we're using the error boundary at the right place in the application, but that's really an integration test and not a unit test.